### PR TITLE
[RLlib] Cleanup examples folder 23: Curiosity (inverse dynamics model based) RLModule example.

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup examples folder 23: Curiosity (inverse dynamics model based) RLModule example.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/learner.py [hunk 1]
```diff
@@ -886,7 +883,7 @@ def compute_loss_for_module(
         self,
         *,
         module_id: ModuleID,
-        config: Optional["AlgorithmConfig"] = None,
+        config: "AlgorithmConfig",
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:886]: Do we now always need to provide a config? I think for most algorithms this is not needed because `self.config` should be available.

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 2]
```diff
@@ -163,13 +163,6 @@ def restore_from_path(self, *args, **kwargs):
     def get_metadata(self, *args, **kwargs):
         self.unwrapped().get_metadata(*args, **kwargs)
 
-    # TODO (sven): Figure out a better way to avoid having to method-spam this wrapper
```
Comment (simonsays1980) [inline rllib/core/rl_module/torch/torch_rl_module.py:166]: Ah nice. This was still there.

File: rllib/examples/curiosity/inverse_dynamics_model_based_curiosity.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                    "FFFFFFFFFFFG",
+                ],
+                "is_slippery": False,
+                # Limit the number of steps the agent is allowed to make in the env to
+                # make it almost impossible to learn without the curriculum.
+                "max_episode_steps": 22,
+            },
+        )
+        # Use our custom `curiosity` method to set up the ICM and our PPO/ICM-Learner.
+        .curiosity(
```
Comment (simonsays1980) [inline rllib/examples/curiosity/inverse_dynamics_model_based_curiosity.py:197]: Nice!

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            model_config_dict={
+                "feature_net_hiddens": self.config.curiosity_feature_net_hiddens,
+                "feature_net_activation": self.config.curiosity_feature_net_activation,
+                "inverse_net_hiddens": self.config.curiosity_inverse_net_hiddens,
+                "inverse_net_activation": self.config.curiosity_inverse_net_activation,
+                "forward_net_hiddens": self.config.curiosity_forward_net_hiddens,
+                "forward_net_activation": self.config.curiosity_forward_net_activation,
+            },
+        )
+        self.add_module(
```
Comment (simonsays1980) [inline rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:88]: Awesome!

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            self.curiosity_eta = curiosity_eta
+        return self
+
+
+class PPOTorchLearnerWithCuriosity(PPOTorchLearner):
+    def build(self):
+        super().build()
+
+        # Assert, we are only training one policy (RLModule).
+        assert len(self.module) == 1 and DEFAULT_MODULE_ID in self.module
```
Comment (simonsays1980) [inline rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:72]: Would we need to set up a single agent's module to be a MultiRLModule to add curiosity to a single agent? 

In this case it might be more handy to provide Curiosity as a wrapper module that wraps existent RLModules into one that also has curiosity.

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            },
+        )
+        self.add_module(
+            module_id=ICM_MODULE_ID,
+            module_spec=icm_spec,
+        )
+
+        # Prepend a "add-NEXT_OBS-from-episodes-to-train-batch" connector piece (right
+        # after the corresponding "add-OBS-..." default piece).
+        if self.config.add_default_connectors_to_learner_pipeline:
```
Comment (simonsays1980) [inline rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:95]: This is so easy now to provide the learner with more functionality.

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        fwd_out: Dict[str, Any],
+        batch: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        # Compute the ICM loss first (so we'll have the chance to change the rewards
+        # in the batch for the "main" RLModule (before we compute its loss with the
+        # intrinsic rewards).
+        icm = self.module[ICM_MODULE_ID]
+        # Send the exact same batch to the ICM module that we used for the "main"
+        # RLModule's forward pass.
+        icm_fwd_out = icm.forward_train(batch=batch[DEFAULT_MODULE_ID])
```
Comment (simonsays1980) [inline rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:113]: Really nice example of how to override the different parts of an algorithm to implement custom logic.

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            learner=self,
+            module_id=ICM_MODULE_ID,
+            config=self.config.get_config_for_module(ICM_MODULE_ID),
+            batch=batch[DEFAULT_MODULE_ID],
+            fwd_out=icm_fwd_out,
+        )
+
+        # Add intrinsic rewards from ICM's `fwd_out` (multiplied by factor `eta`)
+        # to "main" module batch's extrinsic rewards.
+        batch[DEFAULT_MODULE_ID][Columns.REWARDS] += (
```
Comment (simonsays1980) [inline rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:125]: We could also write a connector piece to perform this logic, couldn't we?

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+            print(icm_net.forward_train(input_dict))
+
+            # Print out the number of parameters.
+            num_all_params = sum(int(np.prod(p.size())) for p in icm_net.parameters())
+            print(f"num params = {num_all_params}")
+    """
+
+    @override(TorchRLModule)
+    def setup(self):
+        # Get the IDM achitecture settings from the our RLModuleConfig's (self.config)
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:93]: Small nit: "the our" -> "our"

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        # `model_config_dict` property:
+        cfg = self.config.model_config_dict
+
+        feature_dim = cfg.get("feature_dim", 288)
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:97]: :D How did you come up with this number? Is it in the paper?

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        # Build the feature model (encoder of observations to feature space).
+        layers = []
+        dense_layers = cfg.get("feature_net_hiddens", (256, 256))
+        # `in_size` is the observation space (assume a simple Box(1D)).
+        in_size = self.config.observation_space.shape[0]
+        for out_size in dense_layers:
+            layers.append(nn.Linear(in_size, out_size))
+            if cfg.get("feature_net_activation"):
+                layers.append(
+                    get_activation_fn(cfg["feature_net_activation"], "torch")()
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:108]: Nice use of our help function!

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+                )
+            in_size = out_size
+        # Last feature layer of n nodes (feature dimension).
+        layers.append(nn.Linear(in_size, feature_dim))
+        self._feature_net = nn.Sequential(*layers)
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:113]: Dumb question: did you want to show with this example also hwo to build a Torch module from scratch? We could simply use our predefined MLPs.

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        # `model_config_dict` property:
+        cfg = self.config.model_config_dict
+
+        feature_dim = cfg.get("feature_dim", 288)
+
+        # Build the feature model (encoder of observations to feature space).
+        layers = []
+        dense_layers = cfg.get("feature_net_hiddens", (256, 256))
+        # `in_size` is the observation space (assume a simple Box(1D)).
+        in_size = self.config.observation_space.shape[0]
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:103]: This works only for 1d spaces, does it?

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+                    ).float(),
+                ],
+                dim=-1,
+            )
+        )
+
+        # Forward loss term: Predicted phi - given phi and action - vs actually observed
+        # phi (square-root of L2 norm). Note that this is the intrinsic reward that
+        # will be used and the mean of this is the forward net loss.
+        forward_l2_norm_sqrt = 0.5 * torch.sum(
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:181]: This looks to me like half the MSE loss - is this intended? We could use here for simplification `torch.nn.MSELoss` with `reduction=sum` - and if it should be L2 loss a sqrt on top of it.

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome example. Added some comments where I had questions. Furthermore, if we decide to provide exploration again as a feature we might need wrappers for learners and modules.

**Comment by sven1977** [comment]
> LGTM. Awesome example. Added some comments where I had questions. Furthermore, if we decide to provide exploration again as a feature we might need wrappers for learners and modules.

Great point. I think this example is a first step to making something like "curiosity" pluggable into any algorithm. This same example would work for DQN with only a handful of lines changed (ICM RLModule: unchanged; PPOTorchLearnerWCuriosity: flip over to DQNTorchLearnerWCuriosity w/o really having to change any code; config: stays the same except for flipping out PPOConfig vs DQNConfig).

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py (hunk 14 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:181]: Yeah, correct. It's simple MSE loss. I'm always afraid of using these baked-in loss functions b/c they have no transparency and might do things differently from what my written-out code does :D

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py (hunk 13 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:103]: Currently, yes. If users need e.g. image-obs support, they can now easily write their own ICMs. I didn't want to make the example ICM too complicated.

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py (hunk 12 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:113]: Good point, but yes, I wanted to show a more bare-bones approach w/o using too many RLlib-related utilities. Using only torch makes this RLModule much more readable for new users.

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py (hunk 10 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:97]: yes :D

File: rllib/core/learner/learner.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:886]: I felt like this is the better solution for users. 2 reasons:
* Users normally override `compute_loss_for_module`, so now they do NOT have to implement a logic, where config is None.
* Users do NOT normally override the more top-level `compute_loss`, so we can easily provide each module's individual config through our base implementations.

In other words, if we had left this arg to be optional, every user writing a custom loss function would have had to implement a (not too known) logic on how to get the module's individual config.

File: rllib/examples/learners/classes/curiosity_ppo_torch_learner.py (hunk 8 above)
Comment (sven1977) [reply rllib/examples/learners/classes/curiosity_ppo_torch_learner.py:125]: Great question. We might be able to do this, but only if we make 100% sure that the ICM loss (intrinsic rewards are the individual per-batch-item forward loss values) if available before(!) we call the main RLModule's `compute_loss_for_module` method.

For this to work, we'd probably make the loss computation of the ICM already in its `forward_train` call, which is kind of counterintuitive (the `forward_train` should probably NOT compute any actual losses yet). Nut not sure, this is more a philosophical discussion.
So yes, it would probably work also with a custom connector (instead of overriding `compute_loss`).

File: rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py (hunk 9 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py:93]: fixed
